dashboard/app/models/user.rb

```diff
@@ -1055,12 +1055,12 @@ def next_unpassed_visible_progression_level(script)
     # will be redirected to /congrats.
     return nil if completed?(script)
 
-    visible_sls = visible_script_levels(script).reject {|sl| sl.bonus || sl.level.unplugged? || sl.locked?(self)}
+    visible_sls = visible_script_levels(script).reject {|sl| sl.bonus || sl.level.unplugged?}
 
     sl_level_ids = visible_sls.map(&:level_ids).flatten
 
     # Levels the user made progress in
-    ul_level_ids = user_levels_by_level(script).keys
+    ul_level_ids = user_levels.where(script: script).map(&:level_id)
 
     visible_completed_level_ids = sl_level_ids & ul_level_ids
     visible_incomplete_level_ids = sl_level_ids - ul_level_ids
```

Comment:
Interesting - these are wrong (or inaccurately named) too, I think.  I don't think we have filtered the `ul_level_ids` to only completed levels, it may include attempted levels or UserLevels created because the teacher locked or unlocked a stage.